api/lib/opentelemetry/trace/span_id.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # A class that represents a trace identifier. A valid span identifier is an
+    # 8-byte array with at least one non-zero byte.
+    class SpanId
+      INVALID_ID = 0
+      private_constant :INVALID_ID
+
+      MAX_64BIT_UNSIGNED_INT = (1 << 64) - 1
+      private_constant :MAX_64BIT_UNSIGNED_INT
+
+      def self.generate
+        id = rand(MAX_64BIT_UNSIGNED_INT)
+        return generate if id == INVALID_ID
+
+        new(id)
+      end
+
+      def initialize(id)
+        @id = id
+      end
+
+      # Checks if SpanId is valid
+      #
+      # A valid span identifier is an 8-byte array with at least one non-zero
+      # byte.
+      #
+      # @return [Boolean] true if valid
+      def valid?
+        @id != INVALID_ID
+      end
+
+      def ==(other)
+        @id == other.id
+      end
+
+      protected
+
+      attr_reader :id
+
+      INVALID = new(INVALID_ID)
```

Comment:
IIRC this constant won’t be protected, despite its location under that keyword. My Ruby-fu is probably lacking, tho.

Reply:
Didn't want it to be protected. But it has to be defined after the constructor, otherwise it will fail. Moved it out from protected section to avoid confusion